### PR TITLE
Added a "debugger" and a debug-function, that prints a flying-text ov…

### PR DESCRIPTION
**Pull request by alexaulbach**
…er assigned trains

### CODE REVIEW

**Comment by alexaulbach** [comment]
Resolved some conflicts.

**Comment by Yousei9** [comment]
As I understand it this would print text hovering over the train instead of the console?

What are the benefits of using 470 lines and 3 libs (which i honestly don't quite understand yet) to a simplistic
`if log_level >= 3 then printmsg(trainID.." arrived at ".. stop.entity.backer_name) end`

What if a train debug message is printed hovering over a train on the other side of the map?

**Comment by alexaulbach** [comment]
Moin :)

TL;DR: I think it was a bit too early and your comments are useful for me. Forget this or wait with merging, I have a bit time today and try to make the "debugger" much more useful than yet.

> As I understand it this would print text hovering over the train instead of the console?
Hm. Not exactly.
It is possible to print it like now of course (see dbg.print()). But you can turn it on/off while the game is running and it takes only much more CPU-time, if turned on. 
In general it's an idea to use variable functions, a "code-exchanger", to turn features of code on and off. In this case I called it "debugging", but it could be used for "anything". Feature-flags (https://launchdarkly.com/featureflags.html) are  one case for example.

> Does it really need inspect.lua?

Hm. To be honest I commited inspect.lua by accident. I use it personally to see variable contents and I forgot to remove it before commit. Sorry. But the lib is free (not by me) and you can learn a bit Lua internals.

> What are the benefits of using 470 lines and 2 libs to a simplistic "if log_level...."

Well, it's not finished. :) The advantage is, that in cases that the function is turned off it takes only one function-call. O.K. that's a bit more expensive than just "if log_level then ...", but both methods can be used at the same time. But I would use it only as a simple switch:
`global.debugger and game.print('mytext')
`
Very readable.

Now we exchange it to
`global.debugger and dbg.print(3, 'mytext')
`
The first parameter is the log_level. Still readable. printing is either on or off and I see of course the double functionality with the debug_level, but the sense is to be just fast and if on **then** we can think about the log-level. CPU-friendlyness. And online change of log-level via remote-call.

And how does it look in the end? I used a similar scheme in the current version of RadarTracker https://github.com/alexaulbach/RadarTracker . You can have a look at it and turn the debugging on and then you can see, what's currently going on. It helped me a lot with debugging (so I called it "debugger", but as said this technique could be used for completely different things).

> What if a train debug message is printed hovering over a train on the other side of the map?

Well, many possibilities.
- You can add a switch to switch between printing and flying-text
- You can extend the code to print flying-text-debug only, if it is in range of the player, otherwise it's printed
- You can use the log-file instead.

BTW Log-file: I find the log much more useful than printing to console, cause with many trains you cannot see anything anymore. :)

**Comment by Yousei9** [comment]
I would have held on with the merge until I know exactly what it does and why it does what it does regardless. While this may be inefficient, I want know what my projects do and maintain them by myself if needed.

I like the possibilities of having debug messages over the corresponding stop only when the stop is in view. (the whole code is based on stops not trains, maybe that's where your initial confusion came from)
It would probably need a bit of work to make it look like that and most may turn it off anyway once lamp signaling works.
